:vhs: Record the test suite with VCR

dgb opened the pull request:
Changes the test setup to use VCR on any tests that hit the YouTube API.
Also simplifies the use of authentication credentials, and more
granularly disables some tests. Finally, changes `Request` and `Reports`
to make their retry times stubbable in order to speed up test runs.

dgb committed:
:vhs: Record the test suite with VCR

Changes the test setup to use VCR on any tests that hit the YouTube API.
Also simplifies the use of authentication credentials, and more
granularly disables some tests. Finally, changes `Request` and `Reports`
to make their retry times stubbable in order to speed up test runs.

dgb committed:
Initialize config between tests

kangkyu commented on spec/errors/request_error_spec.rb:
```diff
@@ -1,5 +1,5 @@
 require 'spec_helper'
-require 'yt/errors/request_error'
+require 'yt'
```
why do we need this change?

dgb replied:
I think this was a change for something else I was doing. Doesn't look necessary now. Reverted in ea919e9

kangkyu commented on .rspec:
```diff
@@ -1,6 +1,3 @@
---format documentation
```
I think it's nice to have documentation format, but it's your call

dgb replied:
I don't really like it for normal test runs since it makes it impossible to see where the progress is. You can always pass it if you want to see the doc format too.

kangkyu commented on spec/requests/as_account/playlist_spec.rb:
```diff
@@ -206,7 +206,11 @@
+    let(:id) { test_account.channel.playlists.first.id }
+
+    before do
+      allow(Date).to receive(:today).and_return(Date.new(2020, 2, 5))
```
why do we need this?

dgb replied:
Because this test is exercising a function that (ultimately) calls `Date.today`. We need to fix that so that it's deterministic for VCR (otherwise it'll see a request for a different day and fail).

kangkyu commented on yt.gemspec:
```diff
@@ -1,4 +1,3 @@
-# coding: utf-8
```
what is the difference?

dgb replied:
It's necessary for old Ruby (< 2.0 IIRC), which we should drop support for.

kangkyu replied:
thank you!

dgb committed:
Revert file require change

dgb committed:
Clear API key for non-api tests

dgb committed:
Add note about RSpec

dgb committed:
Re-enable playlist specs

Reenable playlist specs with VCR recordings. Some of the recordings are
a bit fudged in that I added some `sleep` calls since it seems like
playlist-related operations are eventually consistent. This seems OK for
now since the main intent for these tests from now on is to serve to
prevent regressions.

I also removed a test that I couldn't get working that was testing some
quirk with the YouTube API.

dgb committed:
Remove duplicative tests

We don't really need to test this live against the API in both account
and API key contexts.

dgb committed:
Re-enable some video request specs

dgb committed:
Remove wonky video tests

There are several tests that are moreso testing quirks with YouTube's
API that I'm not sure are still applicable, and aren't thorougly
asserting anything (i.e. just checking that an error doesn't happen).
I'm removing these because I'm not sure such granular end-to-end tests
provide much value.

dgb committed:
Re-enable playlist item specs

dgb committed:
Remove duplicative playlist item spec

This spec is already a part of the server spec. We don't need to test
this again with acccount auth.

dgb committed:
Re-enable some channel request specs

dgb committed:
Add more account request specs

dgb committed:
Record last account spec

dgb committed:
Stub out Date in date-sensitive tests
